File: python/src/dexcost/attribution/convert.py
"""Convert durable v1 SDK capture into strict attribution-v2 wire records."""

from __future__ import annotations

import logging
import re
from datetime import timedelta
from decimal import ROUND_HALF_UP, Decimal, InvalidOperation
from typing import Any, cast

from dexcost.attribution.types import (
    ATTRIBUTION_UNIT_BY_METRIC,
    AttributionComponent,
    AttributionCostEvidenceSource,
    AttributionCostEvidenceV2,
    AttributionEventV2,
    AttributionProviderIdentityV2,
    AttributionResourceV2,
    AttributionTaskIngestV1,
    AttributionUsageLineV2,
    AttributionUsageMetric,
)
from dexcost.attribution.validate import validate_attribution_event_v2
from dexcost.models._serde import canonical_decimal, iso_canonical
from dexcost.models.event import Event
from dexcost.models.task import Task

_log = logging.getLogger(__name__)
_GIB = Decimal(1024) ** 3
_TWELVE_PLACES = Decimal("0.000000000001")


def _decimal_detail(details: dict[str, Any], *keys: str) -> Decimal | None:
    for key in keys:
        value = details.get(key)
        if isinstance(value, bool) or value is None:
            continue
        if isinstance(value, (int, float, Decimal, str)):
            try:
                parsed = Decimal(str(value).strip())
                if parsed.is_finite():
                    return parsed
            except (InvalidOperation, ValueError):
                continue
    return None


def _string_detail(details: dict[str, Any], *keys: str) -> str | None:
    for key in keys:
        value = details.get(key)
        if isinstance(value, str) and value.strip():
            return value
    return None


def _canonical_name(value: str | None, fallback: str) -> str:
    normalized = (value or "").strip().lower()
    normalized = re.sub(r"^https?://", "", normalized)
    normalized = re.sub(r"[^a-z0-9._-]+", "_", normalized)
    normalized = re.sub(r"^[_\-.]+|[_\-.]+$", "", normalized)[:128]
    return normalized or fallback


def _positive_quantity(value: object) -> str | None:
    if value is None or isinstance(value, bool):
        return None
    try:
        decimal = value if isinstance(value, Decimal) else Decimal(str(value))
        if not decimal.is_finite() or decimal <= 0:
            return None
        exponent = cast(int, decimal.as_tuple().exponent)
        if exponent < -12:
            decimal = decimal.quantize(_TWELVE_PLACES, rounding=ROUND_HALF_UP)
        if decimal <= 0:
            return None
        return str(canonical_decimal(decimal))
    except (InvalidOperation, ValueError):
        return None


def _usage_line(metric: AttributionUsageMetric, quantity: object) -> AttributionUsageLineV2 | None:
    normalized = _positive_quantity(quantity)
    if normalized is None:
        return None
    return {
        "metric": metric,
        "quantity": normalized,
        "unit": ATTRIBUTION_UNIT_BY_METRIC[metric],
    }


def _compact_usage(
    lines: list[AttributionUsageLineV2 | None],
) -> list[AttributionUsageLineV2]:
    return [line for line in lines if line is not None]


def _provider_for(event: Event) -> AttributionProviderIdentityV2:
    raw = (event.provider or "").lower()
    name = _canonical_name(event.provider, "unknown")
    service = "api"
    if "openai" in raw:
        name, service = "openai", "responses"
    elif "anthropic" in raw:
        name, service = "anthropic", "messages"
    elif "bedrock" in raw:
        name, service = "aws", "bedrock"
    elif "gemini" in raw or raw == "google":
        name, service = "google", "generate_content"
    elif "cohere" in raw:
        name, service = "cohere", "chat"
    elif "vercel" in raw:
        name, service = "vercel", "ai_sdk"
    elif "langchain" in raw:
        name, service = "langchain", "chat"

    if event.event_type != "llm_call":
        billing_model = _string_detail(event.details, "billing_model")
        service_name = event.service_name
        if event.event_type == "compute_cost":
            if billing_model and billing_model.startswith("azure"):
                name = "azure"
            elif billing_model in {"gce", "cloud_functions"} or (
                billing_model and billing_model.startswith("cloud_")
            ):
                name = "google_cloud"
            elif billing_model == "vercel_fluid":
                name = "vercel"
            elif billing_model == "k8s_pod":
                name = "kubernetes"
            elif billing_model in {"lambda", "fargate", "ec2"}:
                name = "aws"
            else:
                name = _canonical_name(event.provider, "runtime")
            service = _canonical_name(billing_model or service_name, "compute")
        elif event.event_type == "gpu_cost":
            name = _canonical_name(
                _string_detail(event.details, "cloud_provider") or event.provider, "runtime"
            )
            service = _canonical_name(billing_model, "gpu")
        elif event.event_type == "network":
            name = _canonical_name(
                _string_detail(event.details, "cloud_provider") or event.provider, "internet"
            )
            service = "egress"
        else:
            raw_service = service_name or "external"
            if raw_service.startswith("mcp:"):
                name = "mcp"
                service = _canonical_name(raw_service[4:], "tool")
            elif "." in raw_service:
                name = _canonical_name(raw_service, "external")
                service = "http_api"
            else:
                name = _canonical_name(event.provider, _canonical_name(raw_service, "external"))
                service = _canonical_name(raw_service, "api")

    provider: AttributionProviderIdentityV2 = {"name": name, "service": service}
    record_id = _string_detail(event.details, "provider_record_id", "request_id", "call_sid")
    region = _string_detail(event.details, "region", "cloud_region")
    if record_id is not None and len(record_id) <= 256:
        provider["record_id"] = record_id
    if region is not None:
        provider["region"] = _canonical_name(region, "unknown")
    return provider


def _resource_for(event: Event) -> AttributionResourceV2 | None:
    if event.model:
        return {"type": "model", "id": event.model[:256]}
    if event.event_type == "gpu_cost":
        sku = _string_detail(event.details, "gpu_sku", "instance_type")
        if sku:
            return {"type": "sku", "id": sku[:256]}
    if event.event_type == "compute_cost":
        instance = _string_detail(event.details, "instance_type", "architecture")
        if instance:
            return {"type": "instance", "id": instance[:256]}
    return None


def _evidence_for(event: Event) -> AttributionCostEvidenceV2 | None:
    amount = _positive_quantity(event.cost_usd)
    if amount is None:
        return None
    source = event.pricing_source
    if source == "provider_response":
        return {
            "amount": amount,
            "currency": "USD",
            "source": "provider_reported",
            "confidence": "exact" if event.cost_confidence == "exact" else "estimated",
        }
    if source in {"manual", "custom", "rate_per_minute"}:
        return {
            "amount": amount,
            "currency": "USD",
            "source": "manual",
            "confidence": cast(Any, event.cost_confidence),
        }
    is_sdk_catalog = source in {"service_catalog", "litellm", "tokencost"} or bool(
        source and source.startswith(("compute_catalog:", "gpu_catalog:", "egress_catalog:"))
    )
    mapped: AttributionCostEvidenceSource | None
    if source == "rate_registry":
        mapped = "sdk_rate_registry"
    elif is_sdk_catalog:
        mapped = "sdk_catalog"
    else:
        mapped = None
    if mapped is None or not event.pricing_version:
        return None
    return {
        "amount": amount,
        "currency": "USD",
        "source": mapped,
        "confidence": cast(
            Any, "computed" if event.cost_confidence == "exact" else event.cost_confidence
        ),
        "pricing_version": event.pricing_version,
    }


def _component_and_usage(
    event: Event,
) -> tuple[AttributionComponent, list[AttributionUsageLineV2], Decimal | None] | None:
    details = event.details
    if event.event_type in {"retry_marker", "gpu_utilization_signal"}:
        return None
    if event.event_type == "llm_call":
        cached = event.cached_tokens or 0
        provider = (event.provider or "").lower()
        cache_counters_are_disjoint = (
            "anthropic" in provider or "bedrock" in provider or provider == "aws"
        )
        input_tokens = event.input_tokens or 0
        if not cache_counters_are_disjoint:
            input_tokens = max(0, input_tokens - cached)
        cache_write = _decimal_detail(details, "cache_creation_input_tokens")
        reasoning = _decimal_detail(details, "reasoning_output_tokens", "reasoning_tokens")
        output_tokens = Decimal(event.output_tokens or 0)
        if reasoning is not None:
            output_tokens = max(Decimal(0), output_tokens - reasoning)
        usage = _compact_usage(
            [
                _usage_line("input_tokens", input_tokens),
                _usage_line("cache_read_input_tokens", cached),
                _usage_line("cache_write_input_tokens", cache_write),
                _usage_line("output_tokens", output_tokens),
                _usage_line("reasoning_output_tokens", reasoning),
            ]
        )
        if not usage:
            usage.append(cast(AttributionUsageLineV2, _usage_line("request_count", 1)))
        return "llm", usage, None
    if event.event_type == "compute_cost":
        duration_ms = _decimal_detail(details, "duration_ms") or Decimal(0)
        duration_seconds = duration_ms / Decimal(1000)
        if duration_seconds == 0:
            duration_seconds = _decimal_detail(details, "wall_clock_seconds") or Decimal(0)
        memory_bytes = _decimal_detail(details, "memory_bytes_limit", "memory_bytes_peak")
        memory_gib_seconds = (
            None if memory_bytes is None else memory_bytes / _GIB * duration_seconds
        )
        return (
            "compute",
            _compact_usage(
                [
                    _usage_line("compute_seconds", duration_seconds),
                    _usage_line("vcpu_seconds", _decimal_detail(details, "vcpu_seconds_used")),
                    _usage_line("memory_gib_seconds", memory_gib_seconds),
                    _usage_line("request_count", _decimal_detail(details, "invocation_count")),
                ]
            ),
            duration_seconds,
        )
    if event.event_type == "gpu_cost":
        duration_seconds = (_decimal_detail(details, "duration_ms") or Decimal(0)) / Decimal(1000)
        measured = _decimal_detail(details, "gpu_seconds_used")
        gpu_count = _decimal_detail(details, "gpu_count") or Decimal(1)
        billing_model = _string_detail(details, "billing_model") or ""
        billed_seconds = (
            measured if billing_model == "per_gpu_second_active" else duration_seconds * gpu_count
        )
        if billed_seconds is None:
            billed_seconds = measured
        return (
            "gpu",
            _compact_usage([_usage_line("gpu_seconds", billed_seconds)]),
            duration_seconds,
        )
    if event.event_type == "network":
        return (
            "network",
            _compact_usage(
                [
                    _usage_line("bytes_out", _decimal_detail(details, "request_bytes")),
                    _usage_line("bytes_in", _decimal_detail(details, "response_bytes")),
                ]
            ),
            None,
        )
    if event.event_type != "external_cost":
        return None
    explicit_quantity = _decimal_detail(details, "attribution_usage_quantity")
    explicit_metric = _string_detail(details, "attribution_usage_metric")
    per = _canonical_name(_string_detail(details, "attribution_usage_per"), "request")
    inferred_metric: AttributionUsageMetric
    if "page" in per:
        inferred_metric = "page_count"
    elif "credit" in per:
        inferred_metric = "credit_count"
    elif "image" in per:
        inferred_metric = "image_count"
    elif "call" in per:
        inferred_metric = "call_count"
    elif "character" in per:
        inferred_metric = "characters"
    else:
        inferred_metric = "request_count"
    metric = (
        cast(AttributionUsageMetric, explicit_metric)
        if explicit_metric in ATTRIBUTION_UNIT_BY_METRIC
        else inferred_metric
    )
    return "external", _compact_usage([_usage_line(metric, explicit_quantity or 1)]), None


def to_attribution_event_v2(event: Event) -> AttributionEventV2 | None:
    """Convert a durable SDK event into a strict, details-free v2 wire event."""
    mapped = _component_and_usage(event)
    if mapped is None:
        return None
    component, usage, duration_seconds = mapped
    if not usage:
        usage.append(cast(AttributionUsageLineV2, _usage_line("request_count", 1)))
    occurred_at = iso_canonical(event.occurred_at)
    converted: AttributionEventV2 = {
        "schema_version": "2",
        "event_id": str(event.event_id),
        "task_id": str(event.task_id),
        "occurred_at": occurred_at,
        "observed_at": occurred_at,
        "component": component,
        "provider": _provider_for(event),
        "lifecycle": {"state": "final", "revision": 1},
        "usage": usage,
    }
    resource = _resource_for(event)
    if resource is not None:
        converted["resource"] = resource
    evidence = _evidence_for(event)
    if evidence is not None:
        converted["cost_evidence"] = evidence
    if event.is_retry and event.retry_of:
        converted["retry_of"] = str(event.retry_of)
    has_time_based_usage = any(line["unit"].endswith("Seconds") for line in usage)
    if has_time_based_usage or (duration_seconds is not None and duration_seconds > 0):
        offset_microseconds = 0
        if duration_seconds is not None and duration_seconds > 0:
            offset_microseconds = int(
                (duration_seconds * Decimal(1_000_000)).to_integral_value(rounding=ROUND_HALF_UP)
            )
        converted["usage_period"] = {
            "start_at": iso_canonical(
                event.occurred_at - timedelta(microseconds=offset_microseconds)
            ),
            "end_at": occurred_at,
        }

    validation = validate_attribution_event_v2(converted)
    if not validation.success:
        _log.warning(
            "Event %s cannot be represented by attribution v2: %s",
            event.event_id,
            ", ".join(issue.path for issue in validation.issues),
        )
        return None
    return converted


def to_attribution_task_ingest_v1(task: Task) -> AttributionTaskIngestV1:
    """Serialize only the task fields accepted by the ingestion boundary."""
    return {
        "task_id": str(task.task_id),
        "task_type": task.task_type,
        "status": cast(Any, task.status),
        "started_at": iso_canonical(task.started_at),
        "ended_at": iso_canonical(task.ended_at) if task.ended_at else None,
        "metadata": cast(dict[str, object], task.metadata),
        "customer_id": task.customer_id,
        "project_id": task.project_id,
        "parent_task_id": str(task.parent_task_id) if task.parent_task_id else None,
        "experiment_id": task.experiment_id,
        "variant": task.variant,
        "schema_version": "1",
    }

File: go/attribution/convert.go
package attribution

import (
	"encoding/json"
	"fmt"
	"log"
	"regexp"
	"strings"
	"time"

	"github.com/DexwoxBusiness/dexcost-sdk/go/core"
	"github.com/shopspring/decimal"
)

var nonCanonicalName = regexp.MustCompile(`[^a-z0-9._-]+`)
var edgePunctuation = regexp.MustCompile(`^[_\-.]+|[_\-.]+$`)
var gib = decimal.NewFromInt(1024).Pow(decimal.NewFromInt(3))

func canonicalTime(value time.Time) string { return value.UTC().Format("2006-01-02T15:04:05.000000Z") }

func decimalDetail(details map[string]interface{}, keys ...string) (decimal.Decimal, bool) {
	for _, key := range keys {
		value, ok := details[key]
		if !ok || value == nil {
			continue
		}
		if b, ok := value.(bool); ok && b {
			continue
		}
		var raw string
		switch v := value.(type) {
		case decimal.Decimal:
			return v, true
		case json.Number:
			raw = v.String()
		case string:
			raw = strings.TrimSpace(v)
		default:
			raw = fmt.Sprint(v)
		}
		parsed, err := decimal.NewFromString(raw)
		if err == nil {
			return parsed, true
		}
	}
	return decimal.Zero, false
}

func stringDetail(details map[string]interface{}, keys ...string) string {
	for _, key := range keys {
		if value, ok := details[key].(string); ok && strings.TrimSpace(value) != "" {
			return value
		}
	}
	return ""
}

func canonicalName(value, fallback string) string {
	value = strings.ToLower(strings.TrimSpace(value))
	value = strings.TrimPrefix(strings.TrimPrefix(value, "https://"), "http://")
	value = nonCanonicalName.ReplaceAllString(value, "_")
	value = edgePunctuation.ReplaceAllString(value, "")
	if len(value) > 128 {
		value = value[:128]
	}
	if value == "" {
		return fallback
	}
	return value
}

func positiveQuantity(value decimal.Decimal) (string, bool) {
	if !value.IsPositive() {
		return "", false
	}
	if -value.Exponent() > 12 {
		value = value.Round(12)
	}
	if !value.IsPositive() {
		return "", false
	}
	return value.String(), true
}

func usageLine(metric UsageMetric, quantity decimal.Decimal) (UsageLineV2, bool) {
	normalized, ok := positiveQuantity(quantity)
	if !ok {
		return UsageLineV2{}, false
	}
	return UsageLineV2{Metric: metric, Quantity: normalized, Unit: UnitByMetric[metric]}, true
}

func appendUsage(lines []UsageLineV2, metric UsageMetric, quantity decimal.Decimal) []UsageLineV2 {
	if line, ok := usageLine(metric, quantity); ok {
		return append(lines, line)
	}
	return lines
}

func providerFor(event core.Event) ProviderIdentityV2 {
	raw := strings.ToLower(event.Provider)
	provider := ProviderIdentityV2{Name: canonicalName(event.Provider, "unknown"), Service: "api"}
	switch {
	case strings.Contains(raw, "openai"):
		provider.Name, provider.Service = "openai", "responses"
	case strings.Contains(raw, "anthropic"):
		provider.Name, provider.Service = "anthropic", "messages"
	case strings.Contains(raw, "bedrock"):
		provider.Name, provider.Service = "aws", "bedrock"
	case strings.Contains(raw, "gemini") || raw == "google":
		provider.Name, provider.Service = "google", "generate_content"
	case strings.Contains(raw, "cohere"):
		provider.Name, provider.Service = "cohere", "chat"
	case strings.Contains(raw, "vercel"):
		provider.Name, provider.Service = "vercel", "ai_sdk"
	case strings.Contains(raw, "langchain"):
		provider.Name, provider.Service = "langchain", "chat"
	}
	if event.EventType != core.EventTypeLLMCall {
		billing := stringDetail(event.Details, "billing_model")
		switch event.EventType {
		case core.EventTypeComputeCost:
			switch {
			case strings.HasPrefix(billing, "azure"):
				provider.Name = "azure"
			case billing == "gce" || billing == "cloud_functions" || strings.HasPrefix(billing, "cloud_"):
				provider.Name = "google_cloud"
			case billing == "vercel_fluid":
				provider.Name = "vercel"
			case billing == "k8s_pod":
				provider.Name = "kubernetes"
			case billing == "lambda" || billing == "fargate" || billing == "ec2":
				provider.Name = "aws"
			default:
				provider.Name = canonicalName(event.Provider, "runtime")
			}
			provider.Service = canonicalName(firstNonEmpty(billing, event.ServiceName), "compute")
		case core.EventTypeGPUCost:
			provider.Name = canonicalName(firstNonEmpty(stringDetail(event.Details, "cloud_provider"), event.Provider), "runtime")
			provider.Service = canonicalName(billing, "gpu")
		case core.EventTypeNetwork:
			provider.Name = canonicalName(firstNonEmpty(stringDetail(event.Details, "cloud_provider"), event.Provider), "internet")
			provider.Service = "egress"
		case core.EventTypeRetryMarker:
			provider.Name, provider.Service = "dexcost", "retry"
		default:
			service := firstNonEmpty(event.ServiceName, "external")
			if strings.HasPrefix(service, "mcp:") {
				provider.Name, provider.Service = "mcp", canonicalName(strings.TrimPrefix(service, "mcp:"), "tool")
			} else if strings.Contains(service, ".") {
				provider.Name, provider.Service = canonicalName(service, "external"), "http_api"
			} else {
				provider.Name, provider.Service = canonicalName(event.Provider, canonicalName(service, "external")), canonicalName(service, "api")
			}
		}
	}
	if recordID := stringDetail(event.Details, "provider_record_id", "request_id", "call_sid"); len(recordID) > 0 && len(recordID) <= 256 {
		provider.RecordID = recordID
	}
	if region := stringDetail(event.Details, "region", "cloud_region"); region != "" {
		provider.Region = canonicalName(region, "unknown")
	}
	return provider
}

func resourceFor(event core.Event) *ResourceV2 {
	if resourceType := stringDetail(event.Details, "attribution_resource_type"); resourceType != "" {
		if resourceID := stringDetail(event.Details, "attribution_resource_id"); resourceID != "" {
			switch resourceType {
			case "model", "sku", "instance", "endpoint", "session", "other":
				return &ResourceV2{Type: resourceType, ID: truncate(resourceID, 256)}
			}
		}
	}
	// Retry markers may also carry model data copied from the failed call.
	// Keep the retry reason as the marker's identity instead of allowing the
	// generic model branch to hide it.
	if event.EventType == core.EventTypeRetryMarker {
		if reason := strings.TrimSpace(event.RetryReason); reason != "" {
			return &ResourceV2{Type: "other", ID: truncate(reason, 256)}
		}
	}
	if event.Model != "" {
		return &ResourceV2{Type: "model", ID: truncate(event.Model, 256)}
	}
	if event.EventType == core.EventTypeGPUCost {
		if sku := stringDetail(event.Details, "gpu_sku", "instance_type"); sku != "" {
			return &ResourceV2{Type: "sku", ID: truncate(sku, 256)}
		}
	}
	if event.EventType == core.EventTypeComputeCost {
		if instance := stringDetail(event.Details, "instance_type", "architecture"); instance != "" {
			return &ResourceV2{Type: "instance", ID: truncate(instance, 256)}
		}
	}
	return nil
}

func evidenceFor(event core.Event) *CostEvidenceV2 {
	amount, ok := positiveQuantity(event.CostUSD)
	if !ok {
		return nil
	}
	if event.EventType == core.EventTypeRetryMarker {
		return &CostEvidenceV2{Amount: amount, Currency: "USD", Source: "manual", Confidence: "exact"}
	}
	source := string(event.PricingSource)
	if source == "" && event.EventType == core.EventTypeNetwork {
		source = stringDetail(event.Details, "egress_pricing_source")
	}
	if source == "provider_response" {
		confidence := "estimated"
		if event.CostConfidence == core.CostConfidenceExact {
			confidence = "exact"
		}
		return &CostEvidenceV2{Amount: amount, Currency: "USD", Source: "provider_reported", Confidence: confidence}
	}
	if source == "manual" || source == "custom" || source == "user_override" {
		return &CostEvidenceV2{Amount: amount, Currency: "USD", Source: "manual", Confidence: string(event.CostConfidence)}
	}
	mapped := ""
	if source == "rate_registry" {
		mapped = "sdk_rate_registry"
	}
	if source == "service_catalog" || source == "litellm" || source == "tokencost" || strings.HasPrefix(source, "compute_catalog:") || strings.HasPrefix(source, "gpu_catalog:") || strings.HasPrefix(source, "egress_catalog:") {
		mapped = "sdk_catalog"
	}
	if mapped == "" || event.PricingVersion == "" {
		return nil
	}
	confidence := string(event.CostConfidence)
	if confidence == "exact" {
		confidence = "computed"
	}
	return &CostEvidenceV2{Amount: amount, Currency: "USD", Source: mapped, Confidence: confidence, PricingVersion: event.PricingVersion}
}

func componentAndUsage(event core.Event) (Component, []UsageLineV2, decimal.Decimal, bool) {
	details := event.Details
	switch event.EventType {
	case core.EventTypeGPUUtilizationSignal:
		return "", nil, decimal.Zero, false
	case core.EventTypeRetryMarker:
		usage := appendUsage(nil, MetricRequestCount, decimal.NewFromInt(1))
		return ComponentExternal, usage, decimal.Zero, true
	case core.EventTypeLLMCall:
		usage := []UsageLineV2{}
		cached := decimal.NewFromInt(int64(intValue(event.CachedTokens)))
		input := decimal.NewFromInt(int64(intValue(event.InputTokens)))
		provider := strings.ToLower(event.Provider)
		if !(strings.Contains(provider, "anthropic") || strings.Contains(provider, "bedrock") || provider == "aws") {
			input = decimal.Max(decimal.Zero, input.Sub(cached))
		}
		cacheWrite, _ := decimalDetail(details, "cache_creation_input_tokens")
		reasoning, _ := decimalDetail(details, "reasoning_output_tokens", "reasoning_tokens")
		output := decimal.NewFromInt(int64(intValue(event.OutputTokens)))
		if reasoning.IsPositive() {
			output = decimal.Max(decimal.Zero, output.Sub(reasoning))
		}
		usage = appendUsage(usage, MetricInputTokens, input)
		usage = appendUsage(usage, MetricCacheReadTokens, cached)
		usage = appendUsage(usage, MetricCacheWriteTokens, cacheWrite)
		usage = appendUsage(usage, MetricOutputTokens, output)
		usage = appendUsage(usage, MetricReasoningTokens, reasoning)
		if len(usage) == 0 {
			usage = appendUsage(usage, MetricRequestCount, decimal.NewFromInt(1))
		}
		return ComponentLLM, usage, decimal.Zero, true
	case core.EventTypeComputeCost:
		durationMS, _ := decimalDetail(details, "duration_ms")
		duration := durationMS.Div(decimal.NewFromInt(1000))
		if duration.IsZero() {
			duration, _ = decimalDetail(details, "wall_clock_seconds")
		}
		memory, memoryOK := decimalDetail(details, "memory_bytes_limit", "memory_bytes_peak")
		usage := []UsageLineV2{}
		usage = appendUsage(usage, MetricComputeSeconds, duration)
		vcpu, _ := decimalDetail(details, "vcpu_seconds_used")
		usage = appendUsage(usage, MetricVCPUSeconds, vcpu)
		if memoryOK {
			usage = appendUsage(usage, MetricMemoryGiBSeconds, memory.Div(gib).Mul(duration))
		}
		invocations, _ := decimalDetail(details, "invocation_count")
		usage = appendUsage(usage, MetricRequestCount, invocations)
		return ComponentCompute, usage, duration, true
	case core.EventTypeGPUCost:
		durationMS, _ := decimalDetail(details, "duration_ms")
		duration := durationMS.Div(decimal.NewFromInt(1000))
		measured, measuredOK := decimalDetail(details, "gpu_seconds_used")
		count, countOK := decimalDetail(details, "gpu_count")
		if !countOK {
			count = decimal.NewFromInt(1)
		}
		billed := duration.Mul(count)
		if stringDetail(details, "billing_model") == "per_gpu_second_active" && measuredOK {
			billed = measured
		} else if billed.IsZero() && measuredOK {
			billed = measured
		}
		return ComponentGPU, appendUsage(nil, MetricGPUSeconds, billed), duration, true
	case core.EventTypeNetwork:
		out, _ := decimalDetail(details, "request_bytes")
		in, _ := decimalDetail(details, "response_bytes")
		usage := appendUsage(nil, MetricBytesOut, out)
		usage = appendUsage(usage, MetricBytesIn, in)
		return ComponentNetwork, usage, decimal.Zero, true
	case core.EventTypeExternalCost:
		quantity, ok := decimalDetail(details, "attribution_usage_quantity")
		if !ok {
			quantity = decimal.NewFromInt(1)
		}
		metric := MetricRequestCount
		explicit := stringDetail(details, "attribution_usage_metric")
		if _, ok := UnitByMetric[UsageMetric(explicit)]; ok {
			metric = UsageMetric(explicit)
		} else {
			per := canonicalName(stringDetail(details, "attribution_usage_per"), "request")
			switch {
			case strings.Contains(per, "page"):
				metric = MetricPageCount
			case strings.Contains(per, "credit"):
				metric = MetricCreditCount
			case strings.Contains(per, "image"):
				metric = MetricImageCount
			case strings.Contains(per, "call"):
				metric = MetricCallCount
			case strings.Contains(per, "character"):
				metric = MetricCharacters
			}
		}
		component := ComponentExternal
		switch stringDetail(details, "attribution_component") {
		case string(ComponentSpeechToText):
			component = ComponentSpeechToText
		case string(ComponentTextToSpeech):
			component = ComponentTextToSpeech
		}
		duration, _ := decimalDetail(details, "attribution_usage_duration_seconds")
		return component, appendUsage(nil, metric, quantity), duration, true
	default:
		return "", nil, decimal.Zero, false
	}
}

// ToEventV2 converts a durable SDK event into a strict, details-free v2 event.
// It returns nil for observability-only or unrepresentable records.
func ToEventV2(event core.Event) *EventV2 {
	component, usage, duration, ok := componentAndUsage(event)
	if !ok {
		return nil
	}
	if len(usage) == 0 {
		usage = appendUsage(usage, MetricRequestCount, decimal.NewFromInt(1))
	}
	occurred := canonicalTime(event.OccurredAt)
	converted := &EventV2{SchemaVersion: "2", EventID: event.EventID.String(), TaskID: event.TaskID.String(), OccurredAt: occurred, ObservedAt: occurred, Component: component, Provider: providerFor(event), Resource: resourceFor(event), Lifecycle: LifecycleV2{State: "final", Revision: 1}, Usage: usage, CostEvidence: evidenceFor(event)}
	if event.IsRetry && event.RetryOf != nil {
		converted.RetryOf = event.RetryOf.String()
	}
	hasTime := false
	for _, line := range usage {
		if strings.HasSuffix(string(line.Unit), "Seconds") {
			hasTime = true
			break
		}
	}
	if hasTime || duration.IsPositive() {
		offset := time.Duration(0)
		if duration.IsPositive() {
			micros := duration.Mul(decimal.NewFromInt(1_000_000)).Round(0).IntPart()
			offset = time.Duration(micros) * time.Microsecond
		}
		converted.UsagePeriod = &UsagePeriodV2{StartAt: canonicalTime(event.OccurredAt.Add(-offset)), EndAt: occurred}
	}
	validation := ValidateEventV2(converted)
	if !validation.Success {
		paths := make([]string, 0, len(validation.Issues))
		for _, issue := range validation.Issues {
			paths = append(paths, issue.Path)
		}
		log.Printf("[dexcost] event %s cannot be represented by attribution v2: %s", event.EventID, strings.Join(paths, ", "))
		return nil
	}
	return converted
}

func ToTaskIngestV1(task core.Task) TaskIngestV1 {
	result := TaskIngestV1{TaskID: task.TaskID.String(), TaskType: task.TaskType, Status: string(task.Status), StartedAt: canonicalTime(task.StartedAt), Metadata: task.Metadata, CustomerID: optionalString(task.CustomerID), ProjectID: optionalString(task.ProjectID), ExperimentID: optionalString(task.ExperimentID), Variant: optionalString(task.Variant), SchemaVersion: "1"}
	if task.EndedAt != nil {
		value := canonicalTime(*task.EndedAt)
		result.EndedAt = &value
	}
	if task.ParentTaskID != nil {
		value := task.ParentTaskID.String()
		result.ParentTaskID = &value
	}
	if result.Metadata == nil {
		result.Metadata = map[string]interface{}{}
	}
	return result
}

func optionalString(value string) *string {
	if value == "" {
		return nil
	}
	return &value
}
func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}
func truncate(value string, max int) string {
	if len(value) > max {
		return value[:max]
	}
	return value
}
func intValue(value *int) int {
	if value == nil {
		return 0
	}
	return *value
}
